quinn-proto/src/connection/mod.rs

```diff
@@ -2376,38 +2401,67 @@ where
                         .retire_cids
                         .extend(retired);
 
-                    // Must be after retire_prior_to processing in case this CID has the highest
-                    // permitted sequence number
+                    let mut valid_rem_cid = true;
+                    if frame.retire_prior_to > self.rem_cid_seq {
+                        // If self.rem_cid_seq is earlier than RETIRE_PRIOR_TO, we must retire the current active remote CID.
+                        // Try to find one valid remote CID that is not retired yet.
+                        match self.retire_rem_cid() {
+                            Some(cid) => self.apply_new_rem_cid(cid),
+                            None => valid_rem_cid = false,
+                        };
+                    }
+
                     use crate::cid_queue::InsertError;
-                    match self.rem_cids.insert(IssuedCid {
+                    let mut new_rem_cid = IssuedCid {
                         sequence: frame.sequence,
                         id: frame.id,
                         reset_token: frame.reset_token,
-                    }) {
-                        Ok(()) => {}
-                        Err(InsertError::ExceedsLimit) => {
-                            return Err(TransportError::CONNECTION_ID_LIMIT_ERROR(""));
-                        }
-                        Err(InsertError::Retired) => {
-                            trace!("discarding already-retired");
-                            self.spaces[SpaceId::Data]
-                                .pending
-                                .retire_cids
-                                .push(frame.sequence);
-                            continue;
+                    };
+
+                    // Guarantee active remote cid (self.rem_cid) always has the smallest sequence number among valid ones in CidQueue
+                    if valid_rem_cid && self.rem_cids.read_offset() == frame.sequence + 1 {
+                        let swap_cid = new_rem_cid;
+                        new_rem_cid.sequence = self.rem_cid_seq;
+                        new_rem_cid.id = self.rem_cid;
+                        new_rem_cid.reset_token = self.peer_params.stateless_reset_token.unwrap();
+                        self.apply_new_rem_cid(swap_cid);
+                    }
```

Comment:
> If we receive NEW_CONNECTION_ID frames out of order, this could lead to an already-used CID being returned to the CID queue. If a migration then occurs and we switch to the queued CID, then that CID will have been used on both sides of the migration, allowing an observed to link the two flows. This weakens privacy.

Good catch. I wasn't thinking about the security requirement